app/views/teachers/TeacherStudentView.coffee

```diff
@@ -0,0 +1,392 @@
+
+    # Use lodash to find the last session for our user, @sessions already sorted by changed date
+    session = _.findLast @sessions.models, (s) => s.get('creator') is @user.id
+
+    return unless session
+
+    # Find course for this level session, for it's name
+    # Level.original is the original id, used for level versioning, and connects levels to level sessions
+    for versionedCourse in @classroom.get('courses') ? []
+      for level in versionedCourse.levels
+        if level.original is session.get('level').original
+          # Found the level for our level session in the classroom versioned courses
+          # Find the full course so we can get it's name
+          course = _.find @courses.models, (c) => c.id is versionedCourse._id
+          break
+
+    # Find level for this level session, for it's name
+    level = @levels.findWhere({original: session.get('level').original})
+
+    # Update last played string based on what we found
+    @lastPlayedString = ""
+    @lastPlayedString += course.get('name') if course
+    @lastPlayedString += ": " if course and level
+    @lastPlayedString += level.get('name') if level
+    @lastPlayedString += ", on " if @lastPlayedString
+    @lastPlayedString += moment(session.get('changed')).format("LLLL")
+    # console.log (moment(session.get('changed')).format("LLLL"))
+    # Rerun template/jade file to display new last played string
+    @render()
+
+  updateLevelProgressMap: ->
+    return unless @courses.loaded and @levels.loaded and @sessions?.loaded and @user?.loaded
+
+    # Map levels to sessions once, so we don't have to search entire session list multiple times below
+    @levelSessionMap = {}
+    # for session in @sessions.models when session.get('creator') is @studentID
+    for session in @sessions.models when session.get('creator') is @studentID
+      @levelSessionMap[session.get('level').original] = session
+
+
+    # Create mapping of level to student progress
+    @levelProgressMap = {}
+    for versionedCourse in @classroom.get('courses') ? []
+      for versionedLevel in versionedCourse.levels
+        session = @levelSessionMap[versionedLevel.original]
+        if session
+          if session.get('state')?.complete
+            @levelProgressMap[versionedLevel.original] = 'complete'
+          else
+            @levelProgressMap[versionedLevel.original] = 'started'
+        else
+          @levelProgressMap[versionedLevel.original] = 'not started'
+
+  updateLevelDataMap: ->
+    return unless @courses.loaded and @levels.loaded and @sessions?.loaded
+
+    @levelData = []
+    for versionedCourse in @classroom.get('courses') ? []
+      course = _.find @courses.models, (c) => c.id is versionedCourse._id
+      for versionedLevel in versionedCourse.levels
+        @playTime = 0 # this and @timesPlayed should probably only count when the levels are completed
+        @timesPlayed = 0
+        @studentTime = 0
```

Comment:
These should be local variables (no `@`) because they're being rewritten many times in the loop, and aren't accessed outside of this function.